### PR TITLE
Remove az_log_set_classifications capability for GA and let the caller filter in the callback.

### PR DESCRIPTION
**Pull request by ahsonkhan**
Fixes https://github.com/Azure/azure-sdk-for-c/issues/1184 and https://github.com/Azure/azure-sdk-for-c/issues/1172

Follow up to https://github.com/Azure/azure-sdk-for-c/issues/1227 and https://github.com/Azure/azure-sdk-for-c/issues/1192

Based on @antkmsft's suggestion, exploring a potentially less disruptive alternative to https://github.com/Azure/azure-sdk-for-c/pull/1317 by limiting changes for GA.

TODO:
- Changelog update
- README update

### CODE REVIEW

File: sdk/inc/azure/core/az_log.h [hunk 1]
```diff
@@ -64,56 +61,32 @@ enum az_log_classification_core
...
-AZ_INLINE void az_log_set_classifications(az_log_classification const classifications[])
-{
-  (void)classifications;
-}
-#endif // AZ_NO_LOGGING
+typedef void (*az_log_message_fn)(az_log_classification classification, az_span message);
 
 /**
- * @brief Sets the function that will be invoked to report an SDK log message.
+ * @brief Set the function that will be invoked to report an SDK log message.
```
Comment (antkmsft) [inline sdk/inc/azure/core/az_log.h:76]: I agree with "Set" vs "Sets", but let's do this change everywhere.

File: sdk/inc/azure/core/az_log.h [hunk 2]
```diff
@@ -64,56 +61,32 @@ enum az_log_classification_core
 };
 
 /**
- * @brief Defines the signature of the callback function that application developers must write in
- * order to receive Azure SDK log messages.
+ * @brief Defines the signature of the callback function that application developers must implement
```
Comment (antkmsft) [inline sdk/inc/azure/core/az_log.h:64]: "must provide"? I had similar feedback in the other PR.

File: sdk/inc/azure/core/internal/az_log_internal.h [hunk 3]
```diff
@@ -11,22 +11,19 @@
 
 #include <azure/core/_az_cfg_prefix.h>
 
-// If the user hasn't registered any classifications, then we log everything.
+#ifdef AZ_NO_LOGGING
```
Comment (antkmsft) [inline sdk/inc/azure/core/internal/az_log_internal.h:14]: Please revert to ifndef, I have more detail in the main PR. Especially here, this is supposed to be a minimalistic change.

File: sdk/samples/storage/blobs/src/blobs_client_example.c [hunk 4]
```diff
@@ -37,14 +37,25 @@ static az_span content_to_upload = AZ_SPAN_LITERAL_FROM_STR("Some test content")
 #endif
 
 // Enable logging
-static void test_log_func(az_log_classification classification, az_span message)
+static bool test_log_should_write_func(az_log_classification classification)
```
Comment (antkmsft) [inline sdk/samples/storage/blobs/src/blobs_client_example.c:40]: In the main PR I had a comment about naming, especially about removing the "test" from the name.

File: sdk/src/azure/core/az_log.c [hunk 5]
```diff
@@ -16,75 +16,43 @@
 
 #ifndef AZ_NO_LOGGING
 
-static az_log_classification const* volatile _az_log_classifications = NULL;
+// Only using volatile here, not for thread safety, but so that the compiler does not optimize what
```
Comment (antkmsft) [inline sdk/src/azure/core/az_log.c:19]: If you want to keep this cmment, place it above, not between volatiles.

File: sdk/src/azure/core/az_log.c [hunk 6]
```diff
@@ -16,75 +16,43 @@
...
-  }
-
-  if (classifications == NULL)
-  {
-    // If the user hasn't registered any classifications, then we log everything.
-    classifications
-        = &classification; // We don't need AZ_LOG_END_OF_LIST to be there, as very first comparison
-                           // is going to succeed and return from the function.
-  }
+  _az_PRECONDITION(classification > 0);
```
Comment (antkmsft) [inline sdk/src/azure/core/az_log.c:31]: Remove this precondition. Especially in this PR, keep the changes to minimum.

File: sdk/src/azure/core/az_log.c [hunk 7]
```diff
@@ -16,75 +16,43 @@
...
+  _az_PRECONDITION_VALID_SPAN(message, 0, true);
+
+  // Copy the volatile fields to local variables so that they don't change within this function.
+  az_log_message_fn const message_callback = _az_log_message_callback;
+
+  // If the user has registered a message_callback, then we log everything.
+  if (_az_log_should_write(classification))
+  {
+    message_callback(classification, message);
```
Comment (antkmsft) [inline sdk/src/azure/core/az_log.c:54]: Bug that you may invoke NULL callback, more details in your main PR.

File: sdk/tests/core/inc/az_test_log.h [hunk 8]
```diff
@@ -6,10 +6,10 @@
 
 // This macro helps with the expected values for tests when verifying logging.
 
-#ifndef AZ_NO_LOGGING
-#define _az_BUILT_WITH_LOGGING(value_if_yes, value_if_no) (value_if_yes)
-#else
+#ifdef AZ_NO_LOGGING
```
Comment (antkmsft) [inline sdk/tests/core/inc/az_test_log.h:9]: Revert

File: sdk/tests/core/test_az_logging.c [hunk 9]
```diff
@@ -66,6 +66,26 @@ static void _log_listener(az_log_classification classification, az_span message)
   }
 }
 
+static bool _log_listener_should_write_request(az_log_classification classification)
```
Comment (antkmsft) [inline sdk/tests/core/test_az_logging.c:69]: I had a bunch of naming suggestions in your main PR, applicable to all tests.

**Review by antkmsft — CHANGES_REQUESTED** [review]
Should we go with this only, please fix the feedback, especially potential NULL callback invocation, plus we need to fix the outdated example in readme.MD (not your fault, butsince we noticed, we should do it), and, more importantly than readme update is let's also remove `az_http_status_code const* status_codes;` from `az_http_policy_retry_options`. Other than that, looks good, but now in addition to this, I like the callback-filter as well.

**Comment by ahsonkhan** [comment]
Closing this in favor of https://github.com/Azure/azure-sdk-for-c/pull/1317

File: sdk/src/azure/core/az_log.c (hunk 5 above)
Comment (ahsonkhan) [reply sdk/src/azure/core/az_log.c:19]: It was above...

**Pull request closed without merging**